### ISSUE TITLE
Presence and frequency penalties are applied to prompt tokens (should be output-only)

### ISSUE THREAD

**Issue by rcfa** [osaurus-ai/vmlx-swift#295]
### Summary

`PresencePenaltyContext` and `FrequencyPenaltyContext` seed their token ring with the **prompt**, so both penalise tokens the model never chose. Under the published definition of these parameters they apply to generated tokens only.

Fix offered as PR #294 (two functions in one file, no API change).

### The code

`Libraries/MLXLMCommon/Evaluate.swift`, identically in both processors:

```swift
mutating public func prompt(_ prompt: MLXArray) {
    ring.loadPrompt(prompt)
}
```

### Why that is wrong

`presence_penalty` and `frequency_penalty` are OpenAI's parameters. HuggingFace has no equivalent, so vLLM — the implementation published values are measured against — is effectively the whole of the specification, and it computes both from the **output** tokens alone:

```python
output_bin_counts, output_mask = get_token_bin_counts_and_mask(
    output_tokens_tensor, vocab_size, num_seqs)
logits -= frequency_penalties.unsqueeze(dim=1) * output_bin_counts
logits -= presence_penalties.unsqueeze(dim=1) * output_mask
```

vLLM's `prompt_tokens_tensor` feeds `repetition_penalties` only.

The divergence is largest exactly where it is least visible. `presenceContextSize` defaults to 20 and prompts are longer than 20 tokens essentially always, so **at the first decode step the window is 100% prompt** — every opening token is penalised for the user's wording. `frequency_penalty` differs in kind rather than degree: the penalty scales with a *count*, so a word the user repeated three times is suppressed three times as hard.

### Why it is worth fixing

Real model cards ship these values. Qwen3 publishes `presence_penalty: 1.5` for non-thinking operation, and the Qwen3.6/3.8 bundles carry it in their sampling modes. A user who copies a published sampling config today gets neither the published behaviour nor a documented alternative.

### Not a bug: `RepetitionContext`

`RepetitionContext` seeds the prompt and **should keep doing so** — HuggingFace defines `repetition_penalty` over `input_ids`, prompt included (`RepetitionPenaltyLogitsProcessor`). The three processors share a shape but not a specification, which is why the fix splits them rather than changing all three.

### Separate, not in the PR: the default window

vLLM applies these penalties over *every* token generated so far — there is no window, because presence is a property of the whole output. The ring here is 20 tokens by default, so a token penalised at step 100 is forgiven by step 120 and penalised again at 140. This needs no code change (`presenceContextSize` is already a caller parameter, and a caller wanting conformance can size the ring to its generation budget), but the **default** may be worth revisiting: 20 is reasonable for `repetitionContextSize`, whose reference is windowed in practice, and less so for a parameter whose reference has no window at all.

**Comment by rcfa** [comment]
Fixed in `main`. Closing — this records which PR settled which part, since the issue raised three
separate things and they landed as three PRs rather than one.

| what the issue raised | resolved by | in `main` as |
|---|---|---|
| penalties seeded from the prompt, against the published output-only definition | #294 | `818e3778d` |
| the 20-token default window, where the reference has no window at all | #296 | `4028b88e0` |
| published sampling configs (e.g. Qwen3's `presence_penalty: 1.5`) not being adopted | #297 | `5a957f71f` |

Verified against `1c17ce8` rather than inferred from the merges: `PresencePenaltyContext.prompt` and
`FrequencyPenaltyContext.prompt` are now no-ops, both context sizes are `Int?` defaulting to `nil`
(unbounded), and `GenerationConfigFile` decodes the two declared penalties.

**`RepetitionContext` still seeds the prompt, deliberately.** That was called out in the issue as *not*
a bug — HuggingFace defines `repetition_penalty` over `input_ids`, prompt included — and the behaviour
is unchanged in `main`. The three processors share a shape but not a specification, which is why the fix
split them instead of changing all three. Worth stating explicitly so nobody later "fixes" it for
consistency.

@jjang-ai also landed a follow-up in #301: the test rewrite pinning the divergence, plus a guard the
optional-size change made necessary. A NON-POSITIVE context size now means disabled — the old guard was
`> 0`, and testing `!= 0` instead would have let a negative slip past into the unbounded branch, turning
"off" into the strongest possible setting in the opposite direction to the caller's intent.

**One note for anyone updating across this**, since it is not purely additive: `presenceContextSize` and
`frequencyContextSize` widened from `Int` (default `20`) to `Int?` (default `nil`). Constructing still
compiles, because `Int` promotes — but code that *reads* either property as a plain `Int` now needs to
unwrap. And the default genuinely changed: a caller who relied on the implicit 20-token window and wants
to keep it should now pass it explicitly.
